finality-aleph/src/justification/requester.rs

```diff
@@ -182,51 +213,64 @@ where
-            let number = *header.number();
-            debug!(target: "aleph-justification", "Trying to request block {:?}", number);
-            self.request_status.save_block_number(number);
-            self.request_status.insert_hash(hash);
-            debug!(target: "aleph-justification", "We have block {:?} with hash {:?}. Requesting justification.", number, header.hash());
-            self.justification_request_scheduler.on_request_sent();
-            self.block_requester.request_justification(&hash, number);
-        } else {
-            debug!(target: "aleph-justification", "Cancelling request, because we don't have block {:?}.", hash);
-        }
+    fn request(&mut self, header: <B as BlockT>::Header) {
+        let number = *header.number();
+        let hash = header.hash();
+        debug!(target: "aleph-justification",
+               "We have block {:?} with hash {:?}. Requesting justification.", number, hash);
+        self.justification_request_scheduler.on_request_sent();
+        self.block_requester.request_justification(&hash, number);
     }
 
     // We request justifications for all the children of last finalized block and a justification
     // for a block of number num on longest branch.
     // Assuming that we request at the same pace that finalization is progressing, the former ensures
-    // that we are up to date with finalization. On the other hand, the former enables fast catch up
+    // that we are up to date with finalization. On the other hand, the latter enables fast catch up
     // if we are behind.
-    // We don't remove the child that it's on the same branch as best since a fork may happen
+    // We also request the child that it's on the same branch as top_wanted since a fork may happen
     // somewhere in between them.
-    fn request_targets(&self, mut top_wanted: NumberFor<B>) -> Vec<<B as BlockT>::Hash> {
-        let blockchain_info = self.backend.info();
+    fn request_targets(&mut self, mut top_wanted: NumberFor<B>) -> Vec<<B as BlockT>::Header> {
+        let blockchain_backend = self.backend.blockchain();
+        let blockchain_info = blockchain_backend.info();
         let finalized_hash = blockchain_info.finalized_hash;
 
-        let mut targets = self.backend.children(finalized_hash).unwrap_or_default();
+        let mut targets = blockchain_backend
+            .children(finalized_hash)
+            .unwrap_or_default()
+            .into_iter()
+            .filter_map(|hash| {
+                if let Ok(Some(header)) = blockchain_backend.header(BlockId::Hash(hash)) {
+                    Some(header)
+                } else {
+                    warn!(target: "aleph-justification",
+                      "Cancelling request for child {:?} of {:?}: no block.", hash, finalized_hash);
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        self.request_status
```

Comment:
This should be only done when `targets.len() > 0`

Reply:
Now it's always true.